aesd-char-driver/main.c

```diff
@@ -29,183 +32,346 @@ struct aesd_dev aesd_device;
-                loff_t *f_pos)
+                  loff_t *f_pos)
 {
-    ssize_t ret = 0, offset = 0;
-    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    struct aesd_buffer_entry *element;
+    struct aesd_dev *a_dev;
+    ssize_t retval = 0;
+    int buffer_count = 0;
+    size_t offset;
+    
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    a_dev = (struct aesd_dev*)filp->private_data;
+
+    if (mutex_lock_interruptible(&a_dev->lock)!=0) {
+	PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+	 mutex_unlock(&a_dev->lock);
+	 return retval;
+    }
 
-    struct aesd_dev* in_dev = filp->private_data;
-	struct aesd_buffer_entry *entry = NULL;
+    element = aesd_circular_buffer_find_entry_offset_for_fpos(&a_dev->circularBuffer, *f_pos, &offset);
+    if(element==NULL) {
+    	 mutex_unlock(&a_dev->lock);
+	 return retval;
+    }
 
-    entry = in_dev->element;
+    if ((element->size - offset) < count)  {
+        *f_pos = *f_pos + (element->size - offset);
+        buffer_count = element->size - offset;
+    } else  {
+        *f_pos = *f_pos + count;
+        buffer_count = count;
+    }
 
-    if (mutex_lock_interruptible(&in_dev->lock) != 0)
+    if (copy_to_user(buf, element->buffptr+offset, buffer_count)) 
     {
-        printk(KERN_INFO "Failed to aquire Mutex.\n");
-        mutex_unlock(&in_dev->lock);
-	    return ret;
+	retval = -EFAULT;
+	mutex_unlock(&a_dev->lock);
+	return retval;
     }
 
-    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&in_dev->circularBuffer, *f_pos, &offset);
+    retval = buffer_count;
 
-    if(entry == NULL)
-    {
-        mutex_unlock(&in_dev->lock);
-        return ret;
-    }
-    else {
-        if(copy_to_user(buf, (entry->buffptr + offset), (entry->size - offset)) == 0) {
-            ret = entry->size - offset;
-        }
-        else {
-            mutex_unlock(&in_dev->lock);
-	        return ret;
-        }
-    }
+    mutex_unlock(&a_dev->lock);
 
-    return ret;
+    return retval;
 }
 
 ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
-                loff_t *f_pos)
+                   loff_t *f_pos)
 {
-    ssize_t retval = -ENOMEM;
-    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
-	const char* lBuffptr = NULL;
+    ssize_t retval = 0;
+    int bytes_to_send = 0, index =0; 
+    int entries = 0, total_bytes = 0; 
+    char *element;
+    const char *freed_buff;
+    struct aesd_buffer_entry w_buffer;
+    struct aesd_dev *a_dev = filp->private_data;
+
+    PDEBUG("write %zu bytes with offset %lld", count, *f_pos);
+    
+    if (mutex_lock_interruptible(&a_dev->lock)!=0) {
+	PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+	return -EFAULT;
+    }
+    element = (char *)kmalloc(count, GFP_KERNEL);
+    if (element == NULL)  {
+        retval = -ENOMEM;
+        mutex_unlock(&a_dev->lock);
+  	return retval;
+    }
+    if (copy_from_user(element, buf, count))  {
```

Comment:
@harshberiwal does this command and kmalloc need to be in your locking interval?  What is your lock locking?